### PR TITLE
[Feature] Enhanced and Optimized TLS

### LINKED ISSUE #3529: Dovecot TLSv [Bug]
### Describe the bug

Good day,

The Dovecot config shows that Dovecot uses at least TLSv1.2, but TLSv1.0 and TLSv1.1 are also used and these cannot be deactivated. Several from the Discord server have also confirmed this, it is not an individual problem.

### Tell us how to replicate the bug

Go to Services on Dovecot
2:Scroll down to /etc/dovecot/conf.d/10-ssl.conf

and then you can see the minimum TLSv1.2 although TLS1.1 and TLS1.0 are also activated

### Which components are affected by this bug?

Mail (Exim, Dovecot)

### Hestia Control Panel Version

v1.7.3

### Operating system

Debian 11

### Log capture

_No response_

### PR DESCRIPTION
**Pull request by myrevery**
<h2 align="center">This is a premium feature sponsored to Hestia.</h2>

<p align="center">Just for love, please enjoy!</p>

<h3 align="center">Main Details</h3>
<h4>For web servers (hestia-nginx nginx)</h4>

Enable TLS 1.3 0-RTT
Enable session resumption via tickets (will be automatically rotate periodically)

<h4>For OpenSSL-dependent (dovecot hestia-nginx nginx proftpd vsftpd)</h4>

Secure: AEAD priority, remove all RSA Kx (full FS), remove all SHA1 HMAC
Performance: AES-128 priority, ChaCha priority (for mobile/ARM devices without AES-NI)
<details>
<summary>View OpenSSL-dependent ciphers (with ECDSA certificate)</summary>
<pre>
Hexcode  KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
--------------------------------------------------------------------------------------
TLSv1.2 (server order -- server prioritizes ChaCha ciphers when preferred by clients)
 xc02b   ECDH 253   AESGCM      128      TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
 xcca9   ECDH 253   ChaCha20    256      TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
 xc02c   ECDH 253   AESGCM      256      TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
 xc023   ECDH 253   AES         128      TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
 xc024   ECDH 253   AES         256      TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
TLSv1.3 (server order -- server prioritizes ChaCha ciphers when preferred by clients)
 x1301   ECDH 253   AESGCM      128      TLS_AES_128_GCM_SHA256
 x1303   ECDH 253   ChaCha20    256      TLS_CHACHA20_POLY1305_SHA256
 x1302   ECDH 253   AESGCM      256      TLS_AES_256_GCM_SHA384
---
via testssl.sh
</pre>
</details>
<details>
<summary>View OpenSSL-dependent ciphers (with RSA certificate)</summary>
<pre>
Hexcode  KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
-------------------------------------------------------------------------------------
TLSv1.2 (server order -- server prioritizes ChaCha ciphers when preferred by clients)
 xc02f   ECDH 253   AESGCM      128      TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
 x9e     DH 4096    AESGCM      128      TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
 xcca8   ECDH 253   ChaCha20    256      TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
 xccaa   DH 4096    ChaCha20    256      TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256
 xc030   ECDH 253   AESGCM      256      TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
 x9f     DH 4096    AESGCM      256      TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
 xc027   ECDH 253   AES         128      TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
 x67     DH 4096    AES         128      TLS_DHE_RSA_WITH_AES_128_CBC_SHA256
 xc028   ECDH 253   AES         256      TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
 x6b     DH 4096    AES         256      TLS_DHE_RSA_WITH_AES_256_CBC_SHA256
TLSv1.3 (server order -- server prioritizes ChaCha ciphers when preferred by clients)
 x1301   ECDH 253   AESGCM      128      TLS_AES_128_GCM_SHA256
 x1303   ECDH 253   ChaCha20    256      TLS_CHACHA20_POLY1305_SHA256
 x1302   ECDH 253   AESGCM      256      TLS_AES_256_GCM_SHA384
---
via testssl.sh
</pre>
</details>

<h4>For GnuTLS-dependent (exim)</h4>

Secure: AEAD priority, remove all RSA Kx (full FS)
Performance: AES-128 priority
<details>
<summary>View GnuTLS-dependent ciphers (with ECDSA certificate)</summary>
<pre>
Hexcode  KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
--------------------------------------------------------------------------------------
TLSv1.2 (server order)
 xc02b   ECDH 256   AESGCM      128      TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
 xc02c   ECDH 256   AESGCM      256      TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
 xcca9   ECDH 256   ChaCha20    256      TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
 xc0ac   ECDH 256   AESCCM      128      TLS_ECDHE_ECDSA_WITH_AES_128_CCM
 xc0ad   ECDH 256   AESCCM      256      TLS_ECDHE_ECDSA_WITH_AES_256_CCM
 xc009   ECDH 256   AES         128      TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
 xc00a   ECDH 256   AES         256      TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
TLSv1.3 (server order)
 x1301   ECDH 256   AESGCM      128      TLS_AES_128_GCM_SHA256
 x1302   ECDH 256   AESGCM      256      TLS_AES_256_GCM_SHA384
 x1303   ECDH 256   ChaCha20    256      TLS_CHACHA20_POLY1305_SHA256
 x1304   ECDH 256   AESCCM      128      TLS_AES_128_CCM_SHA256
---
via testssl.sh
</pre>
</details>
<details>
<summary>View GnuTLS-dependent ciphers (with RSA certificate)</summary>
<pre>
Hexcode  KeyExch.   Encryption  Bits     Cipher Suite Name (IANA/RFC)
------------------------------------------------------------------------------------
TLSv1.2 (server order)
 xc02f   ECDH 256   AESGCM      128      TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
 xc030   ECDH 256   AESGCM      256      TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
 xcca8   ECDH 256   ChaCha20    256      TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
 xc013   ECDH 256   AES         128      TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
 xc014   ECDH 256   AES         256      TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
 x9e     DH 2048    AESGCM      128      TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
 x9f     DH 2048    AESGCM      256      TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
 xccaa   DH 2048    ChaCha20    256      TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256
 xc09e   DH 2048    AESCCM      128      TLS_DHE_RSA_WITH_AES_128_CCM
 xc09f   DH 2048    AESCCM      256      TLS_DHE_RSA_WITH_AES_256_CCM
 x33     DH 2048    AES         128      TLS_DHE_RSA_WITH_AES_128_CBC_SHA
 x39     DH 2048    AES         256      TLS_DHE_RSA_WITH_AES_256_CBC_SHA
TLSv1.3 (server order)
 x1301   ECDH 256   AESGCM      128      TLS_AES_128_GCM_SHA256
 x1302   ECDH 256   AESGCM      256      TLS_AES_256_GCM_SHA384
 x1303   ECDH 256   ChaCha20    256      TLS_CHACHA20_POLY1305_SHA256
 x1304   ECDH 256   AESCCM      128      TLS_AES_128_CCM_SHA256
---
via testssl.sh
</pre>
</details>

<h3 align="center">Attached</h3>
<h4>For web servers (hestia-nginx nginx)</h4>

Added many MIME types to making them compressible
Sync configuration files with mainline to avoid future debug/upgrade pains

<h3 align="center">Early Trial</h3>

For early birds, welcome to test this feature, any feedback would be greatly appreciated!
Don't forget to undo changes when you're done testing.
<pre>bash <(curl -fsL https://myrevery.com/labs/hestia/early-trial-eot.sh)</pre>

### CODE REVIEW

File: src/deb/nginx/nginx.conf [hunk 1]
```diff
@@ -46,48 +47,49 @@ http {
...
-	  application/json application/javascript application/x-javascript application/xml
-	  application/xml+rss application/x-font-ttf image/svg+xml font/opentype;
+	gzip_types                    text/css text/javascript text/js text/plain text/richtext text/shtml text/x-component text/x-java-source text/x-markdown text/x-script text/xml image/bmp image/svg+xml image/vnd.microsoft.icon image/x-icon font/otf font/ttf font/x-woff multipart/bag multipart/mixed application/eot application/font application/font-sfnt application/font-woff application/javascript application/javascript-binast application/json application/ld+json application/manifest+json application/opentype application/otf application/rss+xml application/ttf application/truetype application/vnd.api+json application/vnd.ms-fontobject application/wasm application/xhtml+xml application/xml application/xml+rss application/x-httpd-cgi application/x-javascript application/x-opentype application/x-otf application/x-perl application/x-protobuf application/x-ttf;
 	gzip_proxied                  any;
-	gzip_disable                  "MSIE [1-6]\.";
+	gzip_disable                  "msie6";
```
Comment (jaapmarcus) [inline src/deb/nginx/nginx.conf]: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=867024

Maybe it is time to just remove it... 

Microsoft IE 6 has been dead for way to many years

File: src/rpm/nginx/nginx.conf [hunk 2]
```diff
@@ -46,47 +47,49 @@ http {
...
-	  application/json application/javascript application/x-javascript application/xml
-	  application/xml+rss application/x-font-ttf image/svg+xml font/opentype;
+	gzip_types                    text/css text/javascript text/js text/plain text/richtext text/shtml text/x-component text/x-java-source text/x-markdown text/x-script text/xml image/bmp image/svg+xml image/vnd.microsoft.icon image/x-icon font/otf font/ttf font/x-woff multipart/bag multipart/mixed application/eot application/font application/font-sfnt application/font-woff application/javascript application/javascript-binast application/json application/ld+json application/manifest+json application/opentype application/otf application/rss+xml application/ttf application/truetype application/vnd.api+json application/vnd.ms-fontobject application/wasm application/xhtml+xml application/xml application/xml+rss application/x-httpd-cgi application/x-javascript application/x-opentype application/x-otf application/x-perl application/x-protobuf application/x-ttf;
 	gzip_proxied                  any;
-	gzip_disable                  "MSIE [1-6]\.";
+	gzip_disable                  "msie6";
```
Comment (jaapmarcus) [inline src/rpm/nginx/nginx.conf]: Same here

File: install/deb/nginx/nginx.conf [hunk 3]
```diff
@@ -71,48 +72,42 @@ http {
...
-	  application/json application/javascript application/x-javascript application/xml
-	  application/xml+rss application/x-font-ttf image/svg+xml font/opentype;
+	gzip_types                      text/css text/javascript text/js text/plain text/richtext text/shtml text/x-component text/x-java-source text/x-markdown text/x-script text/xml image/bmp image/svg+xml image/vnd.microsoft.icon image/x-icon font/otf font/ttf font/x-woff multipart/bag multipart/mixed application/eot application/font application/font-sfnt application/font-woff application/javascript application/javascript-binast application/json application/ld+json application/manifest+json application/opentype application/otf application/rss+xml application/ttf application/truetype application/vnd.api+json application/vnd.ms-fontobject application/wasm application/xhtml+xml application/xml application/xml+rss application/x-httpd-cgi application/x-javascript application/x-opentype application/x-otf application/x-perl application/x-protobuf application/x-ttf;
 	gzip_proxied                    any;
-	gzip_disable                    "MSIE [1-6]\.";
-    
-	# Cloudflare ips
+	gzip_disable                    "msie6";
```
Comment (jaapmarcus) [inline install/deb/nginx/nginx.conf]: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=867024

Maybe it is time to just remove it... 

Microsoft IE 6 has been dead for

File: install/rpm/nginx/nginx.conf [hunk 4]
```diff
@@ -14,114 +15,101 @@ events {
...
+	gzip_static                     on;
+	gzip_vary                       on;
+	gzip_comp_level                 6;
+	gzip_min_length                 1024;
+	gzip_buffers                    128 4k;
+	gzip_http_version               1.1;
+	gzip_types                      text/css text/javascript text/js text/plain text/richtext text/shtml text/x-component text/x-java-source text/x-markdown text/x-script text/xml image/bmp image/svg+xml image/vnd.microsoft.icon image/x-icon font/otf font/ttf font/x-woff multipart/bag multipart/mixed application/eot application/font application/font-sfnt application/font-woff application/javascript application/javascript-binast application/json application/ld+json application/manifest+json application/opentype application/otf application/rss+xml application/ttf application/truetype application/vnd.api+json application/vnd.ms-fontobject application/wasm application/xhtml+xml application/xml application/xml+rss application/x-httpd-cgi application/x-javascript application/x-opentype application/x-otf application/x-perl application/x-protobuf application/x-ttf;
+	gzip_proxied                    any;
+	gzip_disable                    "msie6";
```
Comment (jaapmarcus) [inline install/rpm/nginx/nginx.conf]: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=867024

Maybe it is time to just remove it... 

Microsoft IE 6 has been dead for

File: src/deb/nginx/nginx.conf (hunk 1 above)
Comment (myrevery) [reply src/deb/nginx/nginx.conf]: Agree to remove it by default.

I'm going to bed, I'll finish other parts later.

File: func/upgrade.sh [hunk 5]
```diff
@@ -555,14 +555,30 @@ upgrade_b2_tool() {
 	fi
 }
 
+upgrade_nginx_resolver() {
```
Comment (jaapmarcus) [inline func/upgrade.sh]: Is there a reason why we want to change it every update? One time should be enough I think

Comment (myrevery) [reply func/upgrade.sh]: Or only update for hestia-nginx?

Its configuration file will be overwritten after upgrade.

Comment (jaapmarcus) [reply func/upgrade.sh]: hestia-nginx will get over written in that case we should update:
https://github.com/hestiacp/hestiacp/blob/main/src/deb/nginx/postinst

Only how to solve auto updates ... What if happens hestia-nginx is updated before hestia

Comment (myrevery) [reply func/upgrade.sh]: I'll revert this part of the changes first...

**Comment by myrevery** [comment]
@jaapmarcus To avoid issues after release, we need more early testing.
I provided script for early trial, can you invite expert users on Discord and forum to test it?

**Comment by myrevery** [comment]
@jaapmarcus Ready to merge.

Compatibility tests with popular third-party software have all passed.

Installation tests with most popular providers have also passed.

And ready for the upcoming Debian 12.

**Review by jaapmarcus — APPROVED** [review]

**Comment by jaapmarcus** [comment]
@myrevery  

I just noticed in my logs:

sed: -e expression #1, char 9: unterminated `s' command

It is caused by: 
sed -i "s/""$(grep "IMPORTANT: Manual Action Required" "$HESTIA"/data/users/admin/notifications.conf | awk '{print $1}')""/NID='1'/" "$HESTIA"/data/users/admin/notifications.conf

It is probally caused the fact I have run it multiple times :)

**Comment by jaapmarcus** [comment]
That is the issue

**Comment by jaapmarcus** [comment]
I have "fixed" the issue in https://github.com/hestiacp/hestiacp/pull/3622/commits/2315822bb24e246f2c8331eb3e128ec46dbccc00

Limiting to one should be enough

**Comment by myrevery** [comment]
@jaapmarcus Blow you a kiss~:kissing_heart:

Also you may need to check "/etc/ssl/openssl.cnf", if run multiple times, "[hestia_openssl_sect]" will be added repeatedly.

**Comment by jaapmarcus** [comment]
Yes it does...

[hestia_openssl_sect]
Ciphersuites = TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
Options = PrioritizeChaCha
system_default = hestia_openssl_sect

[hestia_openssl_sect]
Ciphersuites = TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
Options = PrioritizeChaCha
system_default = hestia_openssl_sect

[hestia_openssl_sect]
Ciphersuites = TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
Options = PrioritizeChaCha
system_default = hestia_openssl_sect

[hestia_openssl_sect]
Ciphersuites = TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
Options = PrioritizeChaCha
system_default = hestia_openssl_sect

[hestia_openssl_sect]
Ciphersuites = TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
Options = PrioritizeChaCha
system_default = hestia_openssl_sect

[hestia_openssl_sect]
Ciphersuites = TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
Options = PrioritizeChaCha
system_default = hestia_openssl_sect

[hestia_openssl_sect]
Ciphersuites = TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
Options = PrioritizeChaCha
system_default = hestia_openssl_sect

[hestia_openssl_sect]
Ciphersuites = TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
Options = PrioritizeChaCha

**Comment by jaapmarcus** [comment]
Will add an check to it...

**Comment by myrevery** [comment]
Cause I provided script as the "right" way to early trial, so didn't anticipate run the upgrade script multiple times...:man_shrugging:

**Comment by jaapmarcus** [comment]
> Cause I provided script as the "right" way to early trial, so didn't anticipate run the upgrade script multiple times...🤷‍♂️

No problem. I am probably one of the few persons that run the update more then once....

**Pull request merged**